core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -686,259 +725,113 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
-        traverseToInodeInternal(PathUtils.getPathComponents(descendantUri.getPath()), inodeList,
-            nonPersistedInodes, lockList, lockPattern);
-    if (traversalResult.mFound) {
-      return traversalResult.mLockList;
-    } else {
-      throw new InvalidPathException(descendantUri.getPath()
-          + " path not found in traversal starting from " + inodePath.getUri().getPath() + ".");
-    }
+    createdInodes.add(newInode);
+    LOG.debug("createFile: File Created: {} parent: {}", newInode, currentInodeDirectory);
+    return new CreatePathResult(modifiedInodes, createdInodes);
   }
 
   /**
-   * Locks all descendants of a particular {@link LockedInodePath}. Any directory inode
-   * precedes its descendants in the list.
+   * Returns {@link LockedInodePath}s for all descendants of inodePath.
    *
-   * @param inodePath the root {@link LockedInodePath} to retrieve all descendants from
-   * @param lockPattern the lock type to use
-   * @return an {@link InodeLockList} representing the list of all descendants
+   * The descendents are implicitly locked because the current thread holds a write lock on
+   * inodePath, so no new locks are acquired in this method.
+   *
+   * @param inodePath a locked inode path. It must be write-locked
+   * @return all descendants
    */
-  public LockedInodePathList lockDescendants(LockedInodePath inodePath, LockPattern lockPattern) {
-    List<LockedInodePath> inodePathList = new ArrayList<>();
-    lockDescendantsInternal(inodePath, lockPattern, inodePathList);
-    return new LockedInodePathList(inodePathList);
+  public List<LockedInodePath> getImplicitlyLockedDescendants(LockedInodePath inodePath) {
+    Preconditions.checkState(inodePath.getLockPattern().isWrite());
+
+    List<LockedInodePath> descendants = new ArrayList<>();
+    gatherDescendants(inodePath, descendants);
+    return descendants;
   }
 
-  private void lockDescendantsInternal(LockedInodePath inodePath, LockPattern lockPattern,
-      List<LockedInodePath> inodePathList) {
+  private void gatherDescendants(LockedInodePath inodePath, List<LockedInodePath> descendants) {
     InodeView inode = inodePath.getInodeOrNull();
-    if (inode == null || !inode.isDirectory()) {
+    if (inode == null || inode.isFile()) {
       return;
     }
-    InodeDirectoryView inodeDirectory = (InodeDirectoryView) inode;
-    for (InodeView child : inodeDirectory.getChildren()) {
-      LockedInodePath lockedDescendantPath;
+    InodeDirectoryView dir = (InodeDirectoryView) inode;
+    for (InodeView child : dir.getChildren()) {
+      LockedInodePath childPath = null;
       try {
-        lockedDescendantPath =
-            inodePath.createTempPathForExistingChild(child, inodePath.getLockPattern());
-        inodePathList.add(lockedDescendantPath);
-      } catch (InvalidPathException | FileDoesNotExistException e) {
-        // Inode is no longer a child, continue.
+        childPath = inodePath.lockChild(child, LockPattern.WRITE_EDGE);
```

Comment:
I thought `getImplicitlyLockedDescendants` didn't want to lock additional nodes?

Reply:
This is of subtle, but when you extend an already write-locked inodepath, the sub-path doesn't take any locks. It will use `LockedInodePath#traverseWithoutLocking()`